weto/src/main/java/fi/uta/cs/weto/actions/forum/ForumActions.java

```diff
@@ -345,9 +393,82 @@ public String action() throws Exception
+      Tag topicSubscription = new Tag();
+      topicSubscription.setTaggedId(taskId);
+      topicSubscription.setStatus(tag.getId());
+      topicSubscription.setAuthorId(userId);
+      topicSubscription.setType(TagType.FORUM_TOPIC_SUBSCRIPTION.getValue());
+      topicSubscription.setRank(-1);
+      topicSubscription.insert(conn);
+      // Create notification.
+      createNotification(tag);
       return SUCCESS;
     }
 
+    private void createNotification(Tag tag) {
+      Connection masterConnection = getMasterConnection();
+      Connection courseConnection = getCourseConnection();
+
+      try {
+        ArrayList<Tag> forumSubscriptions = Tag.selectByTaggedIdAndRankAndType(courseConnection, getTaskId(), -1, TagType.FORUM_SUBSCRIPTION.getValue());
+
+        // Get the teachers for the course
+        HashSet<Integer> teacherIdSet = new HashSet<>();
+        for (UserTaskView teacher : UserTaskView.selectByTaskIdAndClusterType(courseConnection, getTaskId(), ClusterType.TEACHERS.getValue())) {
+          teacherIdSet.add(teacher.getUserId());
+        }
+
+        // Map values that will be put to the notification template
+        UserAccount author = UserAccount.select1ById(courseConnection, getCourseUserId());
+        HashMap<String, String> valueMap = new HashMap<>();
+        // User name
+        if (teacherIdSet.contains(author.getId())) {
+          valueMap.put("&user;", author.getFirstName() + " " + author.getLastName());
+        } else {
+          valueMap.put("&user;", anonymousName);
+        }
+        // Forum topic
+        JsonObject topicJson = new JsonParser().parse(tag.getText()).getAsJsonObject();
+        valueMap.put("&forumTitle;", topicJson.get("title").toString());
+
+        String notificationMessage = Notification.getMessageFromTemplate(Notification.FORUM_TOPIC, valueMap);
+        String notificationLink = WetoUtilities.getAppBaseUrlFromServlet(getRequest())
+                + "/viewForum.action?taskId=" + getTaskId()
+                + "&tabId=" + getTabId()
+                + "&dbId=" + getDbId();
+
+        CourseImplementation masterCourse = CourseImplementation.select1ByDatabaseIdAndCourseTaskId(masterConnection, getDbId(), getCourseTaskId());
+        for (int i = 0; i < forumSubscriptions.size(); i++) {
+          Integer authorId = forumSubscriptions.get(i).getAuthorId();
+          if (!authorId.equals(getCourseUserId())) {
+            UserAccount user = UserAccount.select1ById(courseConnection, authorId);
+            UserAccount masterUser = UserAccount.select1ByLoginName(masterConnection, user.getLoginName());
+
+            Notification notification = new Notification(masterUser.getId(), masterCourse.getMasterTaskId(), Notification.FORUM_POST, notificationLink);
+            notification.setMessage(notificationMessage);
+            notification.createNotification(masterConnection, courseConnection);
+          }
+        }
```

Comment:
Make code a bit more readable through for-each looping and continue-statement.
Also shouldn't the notification type be FORUM_TOPIC?
```suggestion
        for (Tag subscription : forumSubscriptions) {
          Integer authorId = subscription.getAuthorId();
          if (!authorId.equals(getCourseUserId())) {
            continue;
          }
          
          UserAccount user = UserAccount.select1ById(courseConnection, authorId);
          UserAccount masterUser = UserAccount.select1ByLoginName(masterConnection, user.getLoginName());

          Notification notification = new Notification(masterUser.getId(), masterCourse.getMasterTaskId(), Notification.FORUM_TOPIC, notificationLink);
          notification.setMessage(notificationMessage);
          notification.createNotification(masterConnection, courseConnection);
        }
```